tensorflow/python/ops/array_ops.py

```diff
@@ -2897,7 +2898,7 @@ def matrix_set_diag(
 
 def _constant_if_small(value, shape, dtype, name):
   try:
-    if np.prod(shape) < 1000:
```

Comment:
Do you know what specifically changed in NumPy that broke this?

This may well be a bug in NumPy, but it's hard to say without an isolated reproducer.

Reply:
https://github.com/tensorflow/tensorflow/issues/47691#issuecomment-814255904

Reply:
What exactly is the parameter ```shape``` here? (As in what are an example type and value for ```shape``` that fails with 1.20 but passes with 1.19?)

Reply:
`symbolic Tensor (gru/strided_slice:0)`

Reply:
I put 5 lines of code that will fail with tensorflow 2.4.1 and numpy 1.21.0 at https://github.com/tensorflow/tensorflow/issues/47691#issuecomment-889506604

Reply:
This one line fix is clearly correct, but it looks like this PR cannot be merged because it's changing versions tested in CI (which I agree with maintainers is not necessarily helpful). Why not submit `np.prod` -> `reduce_prod` as a simple one-line fix? The wheels for TensorFlow 2.5.0 on PyPI do not have `<1.19.2`, they only have `~=1.19.2`. So this one-liner fix will unblock users.

Reply:
Also looked into this whole mess, and it looks like there is indeed no bug in NumPy. The only thing that changed is https://github.com/numpy/numpy/pull/19001 made a TensorFlow bug (this incorrect `np.prod` usage) more visible.

Reply:
> Why not submit np.prod -> reduce_prod as a simple one-line fix?

TF maintainers would probably do so if they had any interest in supporting newer versions. But IMHO it is very bad practice to supply a fix but not be able to test it. Nobody can guarantee that this one line fix is the only bug that created incompatibility with NumPy >= 1.20

> The wheels for TensorFlow 2.5.0 on PyPI do not have `<1.19.2`, they only have `~=1.19.2`. So this one-liner fix will unblock users.

Nope. `~=1.19.2` [still blocks](https://www.python.org/dev/peps/pep-0440/#compatible-release) `1.20.*` and `1.21.*`

Reply:
> Nope. `~=1.19.2` [still blocks](https://www.python.org/dev/peps/pep-0440/#compatible-release) `1.20.*` and `1.21.*`

Argh, tricked again by that funky syntax. Sorry about that.

> Nobody can guarantee that this one line fix is the only bug that created incompatibility with NumPy >= 1.20

There's no guarantee, but the other changes in this PR are not all correct or mergeable. CI should indeed use pinned versions, as was already pointed out. You or someone else could submit a separate PR with only a single CI job change that adds a numpy 1.20.3 test. Now you're changing existing CI jobs that were using 1.18.5 before, and will now jump to 1.20.3. Submitting the fix standalone without changing CI shows that no existing tests are breaking.

> TF maintainers would probably do so

I do agree some more attention seems warranted here. I only got here because of a SciPy bug report where someone could not upgrade because of TensorFlow.

Reply:
> You or someone else could submit a separate PR with only a single CI job change that adds a numpy 1.20.3 test.

Have you reviewed the CI setup of TF? With "internal" and "external", "kokoro", "community build" and so on it is totally unclear to me if it is even possible to add an extra job by a regular PR.

Reply:
So i reverted the changes in the ci_build folder. c996d99

Worst case scenario: This PR now "succeeds" with old NumPy versions, is merged, the issue is closed and nobody ever cares again about actually maintaining TF to support NumPy >= 1.20

Reply:
Note that the comments in #50527 speak about exactly this scenario: We are introducing "fixes" but do not test them.

Reply:
It should not close the issue. It just gets the fix that people have been applying by hand merged.

> Have you reviewed the CI setup of TF? With "internal" and "external", "kokoro", "community build" and so on it is totally unclear to me if it is even possible to add an extra job by a regular PR.

yeah it's a bit of a jungle:)

Reply:
I'll make sure that by the end of the year we'll have support for wide range of dependencies.